### DISCUSSION TITLE
How to close the "deep scrub"?

### DISCUSSION THREAD

**Discussion by zhucan** [rook/rook#10617] [Q&A]
if we close the "deep scrub", is there any effects?

**Comment by jerryjedi** [comment]
no impact for a short time, but will cause Inconsistency of replica data for a long time.
How to tune the parameters and schedule of deep scrub need to be considered carefully, depends on your hardware, application workload and business model (7*24 or 5*8). For example, set the "osd scrub priority" to 4 and set a value to "osd scrub load threshold" are important.

**Comment by fmount** [comment]
Hi @zhucan,
disabling deep scrub is something **not** recommended in a Ceph cluster, mostly because  it's synonymous to a file system check for a local filesystem, and is needed to maintain data integrity. (and of course your cluster will be marked  as `HEALTH_WARN`).
As mentioned by @jerryjedi you can tune the scrubbing behavior, but it really depends on the size of the cluster and many other parameters.
In general, the scrub process generates a list of all objects in the PG being scrubbed, locks a chunk of objects (25 by default), and compare the copy on the primary OSD (in the acting set) with copy present on the replicas. The lock process prevents any I/O onto the corresponding object.
While a soft scrub usually checks the object size and its attributes, the deep scrub reads the data and check the checksum(s) to make sure the bits are indeed the same (which means a deep scrub is an I/O intensive operation and may strain the client requests).

From the `ceph cli` you can check the values for each OSD with something like the following:

```
[ceph: root@standalone /]# ceph daemon /var/run/ceph/ceph-osd.0.asok config show | grep -E "(osd_scrub_begin_hour|osd_scrub_end_hour|osd_scrub_max_interval)"
    "osd_scrub_begin_hour": "0",
    "osd_scrub_end_hour": "0",
    "osd_scrub_max_interval": "604800.000000",
```

and set the right time at which the deep-scrub would run (in case you would run it periodically and you would like to have control over that window).

Of course, those parameters can be set at runtime via:

```
ceph tell osd.* injectargs "--osd-scrub-begin-hour X"
ceph tell osd.* injectargs "--osd-scrub-end-hour Y"
```

or within the cluster through the `ceph config set osd <osd_key> <osd_value>` command.
When you're doing day2 operations over the Ceph cluster (a rolling_update, node or OSDs replacement), it would make sense temporarily disabling the deep-scrub process using `ceph osd set nodeep-scrub`: we do that a lot in the OpenStack context during upgrades/updates or many other procedures, but in general this is recommended for a limited period of time when some maintenance on the cluster is performed.

**Reply by zhucan** [reply]
@fmount Thanks, What are the recommended settings in product?

**Reply by fmount** [reply]
I'd say it really depends on the size of your cluster (how many CephStorage nodes, how many OSDs per node).
You can start analyzing the I/O congestion/saturation and tune the provided variables accordingly.
The `ceph pg dump` command is very helpful on this activity, because you can find the oldest deep scrubbing date or filter by date to get the number of PG processed for a given day, then tune the following scrub/deep_scrub parameters using the syntax I mentioned in the previous reply:

~~~
{
    "mds_max_scrub_ops_in_progress": "5",
    "mon_scrub_inject_crc_mismatch": "0.000000",
    "mon_scrub_inject_missing_keys": "0.000000",
    "mon_scrub_interval": "86400",
    "mon_scrub_max_keys": "100",
    "mon_scrub_timeout": "300",
    "mon_warn_pg_not_deep_scrubbed_ratio": "0.750000",
    "mon_warn_pg_not_scrubbed_ratio": "0.500000",
    "osd_debug_deep_scrub_sleep": "0.000000",
    "osd_deep_scrub_interval": "604800.000000",
    "osd_deep_scrub_keys": "1024",
    "osd_deep_scrub_large_omap_object_key_threshold": "200000",
    "osd_deep_scrub_large_omap_object_value_sum_threshold": "1073741824",
    "osd_deep_scrub_randomize_ratio": "0.150000",
    "osd_deep_scrub_stride": "524288",
    "osd_deep_scrub_update_digest_min_age": "7200",
    "osd_max_scrubs": "1",
    "osd_requested_scrub_priority": "120",
    "osd_scrub_auto_repair": "false",
    "osd_scrub_auto_repair_num_errors": "5",
    "osd_scrub_backoff_ratio": "0.660000",
    "osd_scrub_begin_hour": "0",
    "osd_scrub_begin_week_day": "0",
    "osd_scrub_chunk_max": "25",
    "osd_scrub_chunk_min": "5",
    "osd_scrub_cost": "52428800",
    "osd_scrub_during_recovery": "false",
    "osd_scrub_end_hour": "0",
    "osd_scrub_end_week_day": "0",
    "osd_scrub_extended_sleep": "0.000000",
    "osd_scrub_interval_randomize_ratio": "0.500000",
    "osd_scrub_invalid_stats": "true",
    "osd_scrub_load_threshold": "0.500000",
    "osd_scrub_max_interval": "604800.000000",
    "osd_scrub_max_preemptions": "5",
    "osd_scrub_min_interval": "86400.000000",
    "osd_scrub_priority": "5",
    "osd_scrub_sleep": "0.000000",
}
~~~

## Note 
The above is just an example coming from a standalone deployment, but useful to show the list of the involved parameters you can tune according to a previous analysis on the system load and disk performances.
You can run `ceph tell osd.x bench -f plain` to get some insight with OSD and disk performances, then you can act on parameters like `osd_scrub_load_threshold` to stop scrubbing when the system load (loadavg / number of online cpu) is above the given limit (especially during intensive client I/O) or adjust `osd client op priority` to give more priority to client I/O.

**Reply by zhucan** [reply]
@fmount Thanks

**Reply by zhucan** [reply]
@fmount other question is,  the "deep scrub" and "scrub" only can make sure the bits are indeed the same, but can not  avoid the bits are become consistent， is like this?  so if we want to let the scrub to make sure the bits are indeed the same that we can close it?  but we not want to the status of the ceph cluster to  be WARN, is it possible to add a "config value" to controller it?

**Comment by fmount** [comment]
I think it's not a good idea turning it off as it ensures integrity over the `Ceph` pgs. However, that doesn't mean you can tune it as per my previous notes to give priority to the `client I/O` under certain circumstances.
In theory, starting from an healthy cluster, you can set `mon_warn_not_scrubbed": "0"` in combination with  `noscrub` and `nodeep-scrub` options (this should avoid the `HEALTH_WARN`), but AFAIK there's no `config value` to ignore warnings.
However, if you have an in-progress scrub, I'm not sure the parameter above will just be ignored: I remember something from `nautilus` to stop a scrubbing operation [1], but I wouldn't recommend this path.
Regarding the consistency between PGs, the option `osd_scrub_auto_repair` can be set to true and it works according to `osd_scrub_auto_repair_num_errors` [2].

Hope this helps.

[1] https://github.com/ceph/ceph/pull/36292 
[2] https://docs.ceph.com/en/latest/rados/configuration/osd-config-ref/

**Reply by zhucan** [reply]
if there is a config to ignore the warnings is more better.
